modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexProcessor.java

```diff
@@ -218,15 +224,43 @@ public Index createIndexDynamically(
         IndexDefinition definition,
         SchemaIndexCacheVisitor cacheVisitor
     ) {
-        Index idx = createIndex(cctx, factory, definition);
+        IndexFactory dynamicFactory = (gcctx, indexDefinition) -> {
+            Index idx = factory.createIndex(gcctx, indexDefinition);
+
+            // Under lock.
+            if (fillingIdxs == null)
+                fillingIdxs = new HashSet<>();
+            fillingIdxs.add(idx.indexDefinition().idxName());
+
+            return idx;
+        };
+
+        Index idx = createIndex(cctx, dynamicFactory, definition);
```

Comment:
Looks like this call should be in the try-catch block also. For a visitor lambda we can use IndexName from the definition variable.